updateMultigridQuda performance regression sometime after be1766a519995f8c16b9541c300f883efabc3df8

I've observed a (rather substantial) performance regression in `updateMultigridQuda` sometime after commit be1766a519995f8c16b9541c300f883efabc3df8. The earliest  'bad' commit (which I could compile and which ran without erroring out) that I could identify is b50f16a55d2eaae816c00812b3996138c5c4dfa9. 

Running on four nodes of Juwels Booster on a 64c128 lattice at the physical point, calling `updateMultigridQuda` (without refresh iterations):

### good

```
# TM_QUDA: Time for MG_Preconditioner_Setup_Update 1.128518e+00 s level: 4 proc_id: 0 /HMC/cloverdetratio3light:cloverdetratio_derivative/solve_degenerate/invert_eo_degenerate_quda/MG_Preconditioner_Setup_Update
```

[tunecache_be1766a51.tsv.zip](https://github.com/lattice/quda/files/8839700/tunecache_be1766a51.tsv.zip)

When it performs like this, the Update scales reasonably:

| nodes | seconds |
|---------|-------------|
|    4     |     1.1     |
|    8     |     0.62   |
|   16    |     0.42   |
|   32    |     0.21   |

### bad

```
# TM_QUDA: Time for MG_Preconditioner_Setup_Update 5.131241e+00 s level: 4 proc_id: 0 /HMC/cloverdetratio3light:cloverdetratio_derivative/solve_degenerate/invert_eo_degenerate_quda/MG_Preconditioner_Setup_Update
```

[tunecache_b50f16a55.tsv.zip](https://github.com/lattice/quda/files/8839702/tunecache_b50f16a55.tsv.zip)

This poorly performing behaviour doesn't scale. With commit 9bae409b742765563b28939e93fa5c43b03da20b (which happened to be the develop head commit when I started observing this), I have:

| nodes | seconds |
|---------|-------------|
|    4      |      5.1    |
|    8      |      6.6    |
|   16     |     4.8     |

This suggest perhaps a constant factor.

Testing reproduction on a different machine now.

Some more detective work: the regression occurred from one of these pull requests:
* #1260 (@mathiaswagner): merge 345b3335c7b3c8d61812c608730bc2696679ef04
* #1257 (@maddyscientist): merge 2f9d800e47ff80a1d5986fcf20767c561928d4b9
* #1263 (@mathiaswagner): merge 8e245159ebba8febfb8e493d1e3a2e98cfd7fbbb
* #1264 (@weinbe2): merge 0123a9e1d9422ee8b6b3b1a48307b881a6cac53b
* #1228 (@mathiaswagner) (specifically between commits be1766a519995f8c16b9541c300f883efabc3df8 and b50f16a55d2eaae816c00812b3996138c5c4dfa9)

@kostrzewa I know this is a bit of a pain, but could you possibly test the merge points associated with these merges for the first four PRs above?  These should all be stable.  This will help immensely with narrowing this down.

Regarding running the tmLQCD reproducer.  I ran into problems at the linking stage with tmLQCD:

```make
/usr/bin/ld: ./lib/libhmc.a(update_momenta_fg.o): in function `update_momenta_fg':
update_momenta_fg.c:(.text+0x188): undefined reference to `calculate_fg'
collect2: error: ld returned 1 exit status
/install_dir/libmake: *** Waiting for unfinished jobs....
/usr/bin/ld: ./lib/libhmc.a(update_momenta_fg.o): in function `update_momenta_fg':
update_momenta_fg.c:(.text+0x188): undefined reference to `calculate_fg'
collect2: error: ld returned 1 exit status
make: *** [Makefile:136: offline_measurement] Error 1
/usr/bin/ld: ./lib/libhmc.a(update_momenta_fg.o): in function `update_momenta_fg':
update_momenta_fg.c:(.text+0x188): undefined reference to `calculate_fg'
collect2: error: ld returned 1 exit status
make: *** [Makefile:136: hmc_tm] Error 1
/usr/bin/ld: ./lib/libhmc.a(update_momenta_fg.o): in function `update_momenta_fg':
update_momenta_fg.c:(.text+0x188): undefined reference to `calculate_fg'
collect2: error: ld returned 1 exit status
make: *** [Makefile:136: invert] Error 1

--- a/lib/color_spinor_field.cpp
+++ b/lib/color_spinor_field.cpp
@@ -1438,40 +1438,39 @@ namespace quda
               }
             }
           }
+        }
 
-          if (!gdr_recv) {
-            if (!fused_halo_memcpy) {
-              for (int i = 0; i < nDimComms; i++) {
-                if (comm_dim_partitioned(i)) {
-                  if (halo_location[2 * i + 0] == Device && !comm_peer2peer_enabled(0, i)
-                      && // fuse forwards and backwards if possible
-                      halo_location[2 * i + 1] == Device && !comm_peer2peer_enabled(1, i)) {
+        if (!gdr_recv) {
+          if (!fused_halo_memcpy) {
+            for (int i = 0; i < nDimComms; i++) {
+              if (comm_dim_partitioned(i)) {
+                if (halo_location[2 * i + 0] == Device && !comm_peer2peer_enabled(0, i)
+                    && // fuse forwards and backwards if possible
+                    halo_location[2 * i + 1] == Device && !comm_peer2peer_enabled(1, i)) {
+                  qudaMemcpyAsync(from_face_dim_dir_d[bufferIndex][i][0], from_face_dim_dir_h[bufferIndex][i][0],
+                                  2 * ghost_face_bytes_aligned[i], qudaMemcpyHostToDevice, device::get_default_stream());
+                } else {
+                  if (halo_location[2 * i + 0] == Device && !comm_peer2peer_enabled(0, i))
                     qudaMemcpyAsync(from_face_dim_dir_d[bufferIndex][i][0], from_face_dim_dir_h[bufferIndex][i][0],
-                                    2 * ghost_face_bytes_aligned[i], qudaMemcpyHostToDevice,
-                                    device::get_default_stream());
-                  } else {
-                    if (halo_location[2 * i + 0] == Device && !comm_peer2peer_enabled(0, i))
-                      qudaMemcpyAsync(from_face_dim_dir_d[bufferIndex][i][0], from_face_dim_dir_h[bufferIndex][i][0],
-                                      ghost_face_bytes[i], qudaMemcpyHostToDevice, device::get_default_stream());
-                    if (halo_location[2 * i + 1] == Device && !comm_peer2peer_enabled(1, i))
-                      qudaMemcpyAsync(from_face_dim_dir_d[bufferIndex][i][1], from_face_dim_dir_h[bufferIndex][i][1],
-                                      ghost_face_bytes[i], qudaMemcpyHostToDevice, device::get_default_stream());
-                  }
+                                    ghost_face_bytes[i], qudaMemcpyHostToDevice, device::get_default_stream());
+                  if (halo_location[2 * i + 1] == Device && !comm_peer2peer_enabled(1, i))
+                    qudaMemcpyAsync(from_face_dim_dir_d[bufferIndex][i][1], from_face_dim_dir_h[bufferIndex][i][1],
+                                    ghost_face_bytes[i], qudaMemcpyHostToDevice, device::get_default_stream());
                 }
               }
-            } else if (total_bytes && !halo_host) {
-              qudaMemcpyAsync(ghost_recv_buffer_d[bufferIndex], from_face_h[bufferIndex], total_bytes,
-                              qudaMemcpyHostToDevice, device::get_default_stream());
             }
+          } else if (total_bytes && !halo_host) {
+            qudaMemcpyAsync(ghost_recv_buffer_d[bufferIndex], from_face_h[bufferIndex], total_bytes,
+                            qudaMemcpyHostToDevice, device::get_default_stream());
           }
+        }
 
-          // ensure that the p2p sending is completed before returning
-          for (int dim = 0; dim < nDimComms; dim++) {
-            if (!comm_dim_partitioned(dim)) continue;
-            for (int dir = 0; dir < 2; dir++) {
-              if (comm_peer2peer_enabled(dir, dim))
-                qudaStreamWaitEvent(device::get_default_stream(), ipcCopyEvent[bufferIndex][dim][dir], 0);
-            }
+        // ensure that the p2p sending is completed before returning
+        for (int dim = 0; dim < nDimComms; dim++) {
+          if (!comm_dim_partitioned(dim)) continue;
+          for (int dir = 0; dir < 2; dir++) {
+            if (comm_peer2peer_enabled(dir, dim))
+              qudaStreamWaitEvent(device::get_default_stream(), ipcCopyEvent[bufferIndex][dim][dir], 0);
           }
         }
       } else {

